src/mcmini_private.cpp

```diff
@@ -200,59 +212,49 @@ mc_prepare_to_model_check_new_program()
                                            initialTransition);
 }
 
-MC_PROGRAM_TYPE
-mc_run_initial_trace()
+// nextBranchPoint is an out parameter.
+void
+mc_explore_branch(MCOptional<int> *nextBranchPoint)
 {
-  MC_PROGRAM_TYPE program = mc_fork_new_trace_at_main();
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
+  tid_t backtrackThread;
+
+  if (nextBranchPoint == nullptr /* initial branch */) {
+    mc_fork_new_trace();
+    backtrackThread = TID_MAIN_THREAD;
```

Comment:
`MCOptional<>` attempt to act as `std::optional<T>`, the latter only having been introduced in C++17. Better is to pass the optional itself (instead of a pointer to the optional); check the `hasValue()` method; and then react appropriately.

Reply:
I would prefer that the code be easily accessible by general programmers, rather than someone who knows the latest features of C++.  Both `MCOption<>~ and `std::optional<T>` will insert extra assembly code, and will cause GDB to refuse to show the value (because it's hidden in a new class).  Instead of asking people using GDB to write `nextBranchPoint ->MCOptionalXXXint.value`, it's a lot easier in GDB to just type `nextBranchPoint `.

Basically, I'd like to keep our C++ code close the actual pseudo-code, and not add extra layers that will be understood only by C++ mavens.  C++ is maybe the fourth most popular language after C, Java and Python.  And C++ is now facing competition from things like Go and Rust.  But if we stay close enough to C, then anybody will be able to read the code.

Reply:
The point here is that you should either use `MCOptional<>` or `int*`, but don't combine the two. The concept of an optional value should be familiar enough to most students to be accessible in my opinion (see my larger comment below)